candlepin/candlepin
# Maintain order of messages sent to Qpid

When Qpid is down or in flow_stopped, messages that are
received from Artemis are rolled back and enter the
resend on failure loop. When this happens, messages have
the potential to be sent to Qpid out of order due to retry
delays.

This patch shuts down the ClientConsumer that receives the
messages as soon as it is notified that Qpid is in a bad
state. When Qpid comes back up, the session is restarted
and all messages are sent to Qpid in order.

All EventListeners are now required to implement a method
defining whether Qpid is required. If it does, it will be
notified whenever the Qpid connection goes down, or Qpid
is in a bad state.

There are now two MessageHandler implementations that are used
to determine how to handle messages that they receive.

1) EventMessageReceiver is the default message handler. On failure
it will put the message back on the queue and will retry delivery
on a configurable interval.

2) QpidEventMessageReceiver is used for messages that are intended
to be forwarded on to Qpid. If a message does not make it to Qpid
due to a connection issue, the message is put back at the start
of the queue and will be redelivered in the order it was added once
the ClientConsumer is restarted, or candlepin is restarted. If a
message send fails due to any other exception, the message is logged
and discarded.

In order for the retry logic to behave in the way we intend, the
QpidEventMessageHandler needs to on an address that has custom
address settings. In order to acheive this without having to have
the producer send multiple messages in a loop, messages send to the
event.default address will get forwarded to event.qpid via use of
a non-exclusive diverter. Diverted messages will be received with
a different message ID, however, candlepin will log the original
ID along with the new when the message is received.

This PR also merges the ListenerWrapper into the EventReceiver
as the wrapper was not required.

**How To Test**

1. Deploy candlepin with this PR
`deploy -gtaq`

2. Edit the candlepin.conf file to enable the following logging:
`
log4j.logger.org.candlepin.audit.QpidEventMessageReceiver=DEBUG
log4j.logger.org.candlepin.audit.EventMessageReceiver=DEBUG
`
3. Restart tomcat
`
sudo systemctl restart tomcat
`
4. Tail the candlepin log and use grep to isolate the message logs.
`
tail -f /var/log/candlepin/candlepin.log | grep "Message listener org.candlepin.audit.AMQPBusPublisher"
`
5. Register a system to this candlepin instance. You should notice the logs being displayed for a couple messages that were acknowledged.
`
sudo subscription-manager register --user admin --password admin --org admin
`
6. List the available pools for this system and select the pool ID for your favorite.
`
sudo subscription-manager list --avail --pool-only
`
7. Next we want to kick off a subscribe/remove loop. Let it run for a bunch of iterations. In a separate terminal as root run:
`
while true; do subscription-manager remove --all; subscription-manager attach --pool=YOUR_FAV_POOL_ID_GOES_HERE; date; sleep 1; done
`
8. While this is running you should see messages getting processed by the AMQPBusPublisher. The log message shows two message ids - the actual and the original (in brackets). This is due to the message diverting (forwarding) mentioned in the commit message.
`
2018-05-14 10:18:45,849 [thread=Thread-161 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 105934 [105933]:SUCCESS
2018-05-14 10:18:45,852 [thread=Thread-161 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 105936 [105935]:SUCCESS
2018-05-14 10:18:47,575 [thread=Thread-168 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 105952 [105951]:SUCCESS
2018-05-14 10:18:47,585 [thread=Thread-168 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 105954 [105953]:SUCCESS
2018-05-14 10:18:49,562 [thread=Thread-174 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 105976 [105975]:SUCCESS
`
9. Now stop qpidd while leaving the loop running.
`
sudo systemctl stop qpidd
`
10. Wait a few seconds and you'll notice some failures in the logs from above. You'll also notice that while the loop is still running, you'll stop seeing entries appearing in the log matching AMQPBusPublisher. Similar to the following sample.
`
2018-05-14 10:18:57,324 [thread=Thread-199 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 106066 [106065]:SUCCESS
2018-05-14 10:18:59,046 [thread=Thread-206 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 106082 [106081]:SUCCESS
2018-05-14 10:18:59,049 [thread=Thread-206 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 106084 [106083]:SUCCESS
2018-05-14 10:19:10,795 [thread=Thread-211 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 106110 [106109]:FAILED
2018-05-14 10:19:10,796 [thread=Thread-211 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 106112 [106111]:FAILED
2018-05-14 10:19:10,796 [thread=Thread-211 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 106126 [106125]:FAILED
2018-05-14 10:19:10,796 [thread=Thread-211 (ActiveMQ-client-global-threads)] [=, org=, csid=] DEBUG org.candlepin.audit.QpidEventMessageReceiver - Message listener org.candlepin.audit.AMQPBusPublisher@336f5db6 processed message: 106128 [106127]:FAILED
`
11. Wait a few more iterations of the loop and create a few blank lines in your grepped log file terminal and restart qpid.
`
sudo systemctl start qpidd
`
12. You should notice that a bunch more messages get logged and log entries will continue to get added as long as the loop is running. You should stop the loop at this point. The new log entries should start with the first one that had failed and the message IDs should increase and remain in order. 

**Note:** Message IDs will only be in sequence when bursts of messages are received. Don't be alarmed if you see a gap. It is however expected that they remain in order with gaps.
